service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java

```diff
@@ -575,28 +574,28 @@ public PaymentSystemResponse saveFullOrderToDB(OrderResponseDto dto, String uuid
             dto.getBags(), tariffsInfo);
         checkIfUserHaveEnoughPoints(currentUser.getCurrentPoints(), dto.getPointsToUse());
         long sumToPayInCoins = reduceOrderSumDueToUsedPoints(sumToPayWithoutDiscountInCoins, dto.getPointsToUse());
-
         Order order = isExistOrder(dto, orderId);
         if (orderId != null) {
             checkIsOrderOfCurrentUser(currentUser, order);
         }
         order.setTariffsInfo(tariffsInfo);
         Set<Certificate> orderCertificates = new HashSet<>();
         sumToPayInCoins = formCertificatesToBeSavedAndCalculateOrderSum(dto, orderCertificates, order, sumToPayInCoins);
-
+        if (sumToPayInCoins <= 0) {
+            dto.setShouldBePaid(false);
+        }
         UBSuser userData =
             formUserDataToBeSaved(dto.getPersonalData(), dto.getAddressId(), dto.getLocationId(), currentUser);
-
         getOrder(dto, currentUser, bagsOrdered, sumToPayInCoins, order, orderCertificates, userData);
         eventService.save(OrderHistory.ORDER_FORMED, OrderHistory.CLIENT, order);
-        PaymentSystemResponse paymentSystemResponse = processPayment(dto, order, sumToPayInCoins, currentUser);
+        PaymentSystemResponse paymentSystemResponse;
+        if (sumToPayInCoins > 0 || dto.isShouldBePaid()) {
```

Comment:
you can check `dto.isShouldBePaid()` only because it is always false if `sumToPayInCoins <= 0`
you set it here http://github.com/ita-social-projects/GreenCityUBS/pull/1621/files#diff-d5ba7e4400c6d3e88dfdd27f144c7c964c8819955aa261014b596806cd6b7ac9R584

Reply:
please add tests for the new logic